src/target/mspm0.c

```diff
@@ -0,0 +1,396 @@
+	if (manufacturer != JEP106_MANUFACTURER_TEXAS)
+		return false;
+
+	uint32_t partnum = (deviceid & MSPM0_DEVICEID_PARTNUM_MASK) >> MSPM0_DEVICEID_PARTNUM_SHIFT;
+	if (partnum != TI_DEVID_MSPM0C && partnum != TI_DEVID_MSPM0L && partnum != TI_DEVID_MSPM0L_1227_2228 &&
+		partnum != TI_DEVID_MSPM0G)
+		return false;
+
+	target->driver = "MSPM0";
+	target->target_options |= TOPT_INHIBIT_NRST;
+	target->mass_erase = mspm0_mass_erase;
+
+	uint32_t const sramflash = target_mem32_read32(target, MSPM0_FACTORYREGION_SRAMFLASH);
+	const uint32_t mainflash_size = 1024U *
+		((sramflash & MSPM0_FACTORYREGION_SRAMFLASH_MAINFLASH_SZ_MASK) >>
+			MSPM0_FACTORYREGION_SRAMFLASH_MAINFLASH_SZ_SHIFT);
+	const uint32_t main_num_banks = (sramflash & MSPM0_FACTORYREGION_SRAMFLASH_MAINNUMBANKS_MASK) >>
+		MSPM0_FACTORYREGION_SRAMFLASH_MAINNUMBANKS_SHIFT;
+	const uint32_t sram_size = 1024U *
+		((sramflash & MSPM0_FACTORYREGION_SRAMFLASH_SRAM_SZ_MASK) >> MSPM0_FACTORYREGION_SRAMFLASH_SRAM_SZ_SHIFT);
+	const uint32_t dataflash_size = 1024U *
+		((sramflash & MSPM0_FACTORYREGION_SRAMFLASH_DATAFLASH_SZ_MASK) >>
+			MSPM0_FACTORYREGION_SRAMFLASH_DATAFLASH_SZ_SHIFT);
+
+	target_add_ram32(target, MSPM0_SRAM_BASE, sram_size);
+	mspm0_add_flash(target, MSPM0_FLASH_MAIN, mainflash_size, main_num_banks);
+	if (dataflash_size != 0)
+		mspm0_add_flash(target, MSPM0_FLASH_DATA, dataflash_size, 1U);
+
+#if MSPM0_CONFIG_FLASH_DUMP_SUPPORT
+	target_add_commands(target, mspm0_cmds_list, "MSPM0");
+#endif
+
+	return true;
+}
+
+/* Wait for flash command to finish and return the status word or UINT32_MAX if timout */
+static uint32_t mspm0_flash_wait_done(target_s *const target)
+{
+	platform_timeout_s timeout;
+	platform_timeout_set(&timeout, 500U);
+
+	uint32_t statcmd;
+	while (true) {
+		statcmd = target_mem32_read32(target, MSPM0_FLASHCTL_STATCMD);
+		if (statcmd & MSPM0_FLASHCTL_STAT_DONE)
+			break;
+		if (platform_timeout_is_expired(&timeout))
+			return UINT32_MAX;
+	};
+
+	return statcmd;
+}
+
+static void mspm0_flash_unprotect(target_flash_s *const target_flash)
+{
+	target_mem32_write32(target_flash->t, MSPM0_FLASHCTL_CMDWEPROTA, 0U);
+	target_mem32_write32(target_flash->t, MSPM0_FLASHCTL_CMDWEPROTB, 0U);
+	target_mem32_write32(target_flash->t, MSPM0_FLASHCTL_CMDWEPROTC, 0U);
+}
+
+static void mspm0_flash_unprotect_sector(target_flash_s *const target_flash, const target_addr_t addr)
+{
+	mspm0_flash_s *mspm0_flash = (mspm0_flash_s *)target_flash;
+	unsigned sector = (addr - target_flash->start) / MSPM0_FLASH_SECTOR_SZ;
+
+	if (sector < 32U) {
+		uint32_t mask = ~(1U << sector);
+		target_mem32_write32(target_flash->t, MSPM0_FLASHCTL_CMDWEPROTA, mask);
+	} else if (sector < 256U) {
+		/* Depending whether main flash is multibank or not, the first 4 bits of PROTB either
+		start at 32k or overlay over PROTA responsible range. */
+		unsigned start_protb_sector = mspm0_flash->banks <= 1 ? 32U : 0U;
```

Comment:
Can you please explain how a possible conversion (from what into what) could happen in the next line?

Reply:
While I will change to uint32_t for the peace of mind, I do not see how the code could cause any issues. Both operands for subtraction are `unsigned int` so no integer promotion will happen as they are unsigned `ints` already. I guess they key point here is that the expression result is used as a shift amount value so any type ripple stops right there and does not affect `uint32_t mask`.

Reply:
The `uint32_t` argument is `unsigned long` in this case, which is a different length when `int` is 16-bit on a platform, causing promotion which promotes to *signed* (your `unsigned` becomes `long`). Once that happens, the shifting is all done on *signed* numbers and you open the calculation to all of the pitfalls mentioned above. Remember that just because it's okay on one platform does not mean it's okay on all, because this is a platform-dependant issue.

Reply:
I guess what I fail to see is how the type of the shift amount can influence the resulting type of the shift operation (provided that the shift-by value calculates to a correct one). While the current type of `sector` is indeed problematic and should be kept as `target_addr_t`, or is uint32_t good enough?

Reply:
`uint32_t` is sufficient. Because the target is a 32-bit target, that particular narrowing won't cause problems, and narrowing never signed-unsigned converts.